### PR TITLE
Add doc comment to Pipeline

### PR DESCRIPTION
**Pull request by cbracken**
Adds more detailed documentation to the Pipeline class. Also fixes a typo in fml::Semaphore docs.

No tests since this patch makes no semantic changes.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/common/pipeline.h [hunk 1]
```diff
@@ -36,6 +36,20 @@ size_t GetNextPipelineTraceID();
...
+/// Pipelines support two key operations: produce and consume.
+///
+/// The consumer calls `Consume` to wait for a resource to be produced and
+/// consume it when ready.
+///
+/// The producer calls `Produce` to generate a `ProducerContinuation` which
+/// provides a means to enqueue a resource in the pipeline. When the resource
+/// has been prepared, the producer calls `Complete` on the continuation, which
+/// enqueues the resource and signals the waiting consumer.
+///
```
Comment (goderbauer) [inline shell/common/pipeline.h:49]: Would it make sense to document this on the methods themselves as well?

**Review by goderbauer — APPROVED** [review]
LGTM

File: shell/common/pipeline.h (hunk 1 above)
Comment (cbracken) [reply shell/common/pipeline.h:49]: Done.

**Review by loic-sharma — APPROVED** [review]

**Review by justinmc — APPROVED** [review]
The doc comment makes sense to me as someone not familiar with this pipeline :+1: .

It might not be worth addressing this in these docs, but after reading this I don't know why or where we might use this pipeline.

**Comment by cbracken** [comment]
@justinmc at present (unless someone's added another use) there's actually only a single use of this pipeline, which is the LayerTree pipeline used in the animator/rasterizer. Agreed it would probably be a little clearer if we just hardcoded that in rather than having a generic type, though the generic type sure makes this a lot easier to test :)

I always feel a bit gross about adding "where this is used" type comments but given I don't see this changing anytime soon, maybe I should just add a breadcrumb.

**Comment by cbracken** [comment]
Done.

**Comment by justinmc** [comment]
Sounds good.  I agree about "where this is used" comments but the sentence you added seems pretty useful.  Thanks!

**Pull request merged**